programs/rodeos/streams/rabbitmq.hpp

```diff
@@ -15,28 +15,53 @@ class rabbitmq : public stream_handler {
    std::shared_ptr<rabbitmq_handler>    handler_;
    std::shared_ptr<AMQP::TcpConnection> connection_;
    std::shared_ptr<AMQP::TcpChannel>    channel_;
-   std::string                          name_;
+   std::string                          exchangeName_;
+   std::string                          queueName_;
    std::vector<eosio::name>             routes_;
 
  public:
-   rabbitmq(boost::asio::io_service& io_service, std::vector<eosio::name> routes, std::string address, std::string name)
-       : name_(std::move(name)), routes_(std::move(routes)) {
+   rabbitmq(boost::asio::io_service& io_service, std::vector<eosio::name> routes, std::string address, std::string queue_name)
+       : queueName_(std::move(queue_name)), routes_(std::move(routes)) {
       AMQP::Address amqp_address(address);
-      ilog("Connecting to RabbitMQ address ${a} - Queue: ${q}...", ("a", std::string(amqp_address))("q", name_));
+      ilog("Connecting to RabbitMQ address ${a} - Queue: ${q}...", ("a", std::string(amqp_address))("q", queueName_));
 
-      handler_    = std::make_shared<rabbitmq_handler>(io_service);
-      connection_ = std::make_shared<AMQP::TcpConnection>(handler_.get(), amqp_address);
-      channel_    = std::make_shared<AMQP::TcpChannel>(connection_.get());
+      init(io_service, amqp_address);
+
+      exchangeName_ = DEFAULT_EXCHANGE;
       declare_queue();
    }
 
+   rabbitmq(boost::asio::io_service& io_service, std::vector<eosio::name> routes, std::string address, std::string exchange_name, std::string exchange_type)
+       : exchangeName_(std::move(exchange_name)), routes_(std::move(routes)) {
+      AMQP::Address amqp_address(address);
+      ilog("Connecting to RabbitMQ address ${a} - Exchange: ${e}...", ("a", std::string(amqp_address))("e", exchangeName_));
+
+      init(io_service, amqp_address);
+
+      declare_exchange(exchange_type);
+   }
+
    const std::vector<eosio::name>& get_routes() const override { return routes_; }
 
-   void publish(const char* data, uint64_t data_size) override { channel_->publish("", name_, data, data_size, 0); }
+   void publish(const char* data, uint64_t data_size, const eosio::name& routing_key) override {
+      if (DEFAULT_EXCHANGE == exchangeName_) {
+         channel_->publish(exchangeName_, queueName_, data, data_size, 0);
+      } else {
+         channel_->publish(exchangeName_, routing_key.to_string(), data, data_size, 0);
```

Comment:
I feel like you should store as string in our constructor instead of converting from name -> string for every message that is being published

Reply:
ah very good point, will do

Reply:
ohhhh.... nevermind it actually comes from the filter